### ISSUE TITLE
[Android] No such file or directory when starting a puzzle

### ISSUE THREAD

**Issue by mueller-ma** [chrisboyle/sgtpuzzles#409]
Hi,
I downloaded the apk from fdroid and when starting a puzzle I get this message:
```
Error
/data/app/name.boyle.chris.sgtpuzzles-1/lib/arm/libpuzzlesgen-with-pie.so: open failes: ENOENT (No such file or directory)
```

Do I have to install additional packages?

**Comment by chrisboyle** [comment]
No, this is a file included in the APK, unpacked by the Android OS, but in your case and on one other person's device, Puzzles seems confused as to where to find it. :-(

Please comment with your device's model number and version of Android?

Background for those interested:

This is a failure during [`startGameGenProcess`](https://github.com/chrisboyle/sgtpuzzles/blob/master/app/src/main/java/name/boyle/chris/sgtpuzzles/GamePlay.java#L954): the idea is the OS has unpacked native libraries to a particular place, including the `puzzlesgen` executables, and it should tell me where when I look in [`getApplicationInfo().nativeLibraryDir`](https://developer.android.com/reference/android/content/pm/ApplicationInfo.html#nativeLibraryDir).

The weird thing is that "arm" in the path which doesn't look like the right name. Here's [the list of ABIs](https://developer.android.com/ndk/guides/abis.html), which the APK includes versions for all 7 of; "arm" is closest to "armeabi", which I include.

**Comment by chrisboyle** [comment]
The other user's info:

`Puzzles v2016-12-25-0849-0b34887 - on F8331 / original / Sony/F8331/F8331:6.0.1/39.0.A.1.250/550366403:user/release-keys`

**Comment by chrisboyle** [comment]
Ooooh, I just saw the "fdroid" part of your report. Most likely they have a bad build then. Try [this instead](https://chris.boyle.name/download/SGTPuzzles-2016-12-25-0849-0b34887.apk), and I'll take it up with them.

**Comment by chrisboyle** [comment]
[Reported to F-Droid here](https://gitlab.com/fdroid/fdroiddata/issues/572).

**Comment by mueller-ma** [comment]
Puzzles v2016-12-25-0849-0b34887
Android 6.0.1
Oneplus X
F-Droid

Idk what the other stuff in someone others info is.

**Comment by est31** [comment]
@mueller-ma see my comment in the f-droid issue. We are aware and trying to fix it. I suppose the earlier versions of the app are not affected?

**Comment by mueller-ma** [comment]
@est31 I will try

**Comment by mueller-ma** [comment]
Version 2016-01-13-0847-1d9d6cb and 2015-09-10-2219-eda5a86 don't have this problem

**Comment by est31** [comment]
Thanks for the info!

**Comment by collectorgeneral** [comment]
Was about to open an issue, but OP already opened it. 
Just for information, thanks @chrisboyle , the f-droid one did not work but the APK you linked worked.

**Comment by crogonint** [comment]
FYI Chris, your apk fixed the problem. :)

**Comment by est31** [comment]
can you check the latest APK in f-droid whether the issue is fixed?

**Comment by mueller-ma** [comment]
@est31 @crogonint Still the same problem

**Comment by chrisboyle** [comment]
@est31 I agree, the APK available from https://f-droid.org/repository/browse/?fdid=name.boyle.chris.sgtpuzzles&fdpage=17&fdstyle=grid has the same problem. You can tell because, opening it as a zip, the `lib/*` directories each contain only `libpuzzles.so` where there should also be 2 `libpuzzlesgen*.so` in each such directory.

**Comment by est31** [comment]
Oh, sorry again. Will disable again and work on a fix.

**Comment by crogonint** [comment]
Nope, still doing the same thing. I don't see a newer apk, just the same
one. Are they planning on rebuilding the broken version, our building a new
one?

**Comment by est31** [comment]
@crogonint it was disabled. However the change wont take effect immediately, it always resyncs once per day.

**Comment by crogonint** [comment]
Well that's annoying. I'll backtrack and reinstall your apk, try again
tomorrow. ;)

**Comment by h4711** [comment]
I'm having a the same problem doing my own build of master or tags/2016-12-25-0849-0b34887.
This is on Moto G 2014 with Android 5.
On Vernee Thor with Android 6 the requested path is arm64 instead of arm, but it's missing too.

`$ unzip -l app-debug.apk | grep  libpuzzles
  1642328  1980-00-00 00:00   lib/arm64-v8a/libpuzzles.so
  1006596  1980-00-00 00:00   lib/armeabi-v7a/libpuzzles.so
  1031164  1980-00-00 00:00   lib/armeabi/libpuzzles.so
  2063472  1980-00-00 00:00   lib/mips/libpuzzles.so
  1992224  1980-00-00 00:00   lib/mips64/libpuzzles.so
  1334508  1980-00-00 00:00   lib/x86/libpuzzles.so
  1450448  1980-00-00 00:00   lib/x86_64/libpuzzles.so
`

**Comment by SteffenBauer** [comment]
I had the same problem, then I remembered this line from README-android:
```
IMPORTANT: run the gradle task "buildGameGenerationExecutableDebug", ideally
before installation (or force a rebuild after). This gives you the executable
required to generate each new game ("puzzlesgen").
```
After running this task in Android Studio, puzzles started successfully in the emulator (didn't try so far on a real device).

**Comment by h4711** [comment]
doing a _buildGameGenerationExecutableDebug_ before _build_ makes the missing files available
so _build_ should depend on _buildGameGenerationExecutableDebug_ and that probably will fix F-Droid's version too
Remark: Didn't test on device so far.
Remark: _buildGameGenerationExecutableDebug_ should support builds w/o local.properties

**Comment by chrisboyle** [comment]
I know it's annoying that this separate task is necessary and not invoked automatically. Unfortunately I'm limited by the build system. The limitation is [mentioned](https://sites.google.com/a/android.com/tools/tech-docs/new-build-system/gradle-experimental#TOC-Signing-Config) in the gradle-experimental docs (it says "The DSL for modifying variants and their tasks is very, very limited right now.")

The previous ("stable" gradle) build system became unsuitable because I need to compile the native library (libpuzzles.so) for an earlier target platform than the newest supported one, in order to maintain compatibility.

Despite this it may be possible with some gradle trickery to trigger the extra task at the right time. I'm looking into splitting both the library and the executable out as separate gradle projects, which may make it easier to introduce the correct dependencies between them and the main project. That work is the reason I'm leaving this bug open currently. I would accept a patch that does this while remaining with gradle-experimental. I *might* accept a patch that achieves it by switching to CMake, depending on whether that means all builders need to install CMake (separately from the SDK/NDK).

**Comment by h4711** [comment]
Is there any chance to test for the missing files _during_ build and bail out or at least write some warning?

**Comment by chrisboyle** [comment]
No, not until runtime, because the hard part is to insert *any* task into the normal build process. Inserting a task to check that another task wasn't inserted is not any easier.

I am going to add a runtime check so that people like F-Droid can see immediately on app startup that they have a broken build.

**Comment by h4711** [comment]
... and it will help developers starting to work on puzzles. Please add some hint to the dialog, where to find build instructions. You could also add a build.sh to the source that calls ./gradlew, so everybody would see the need for the extra call at first.

This issue is resolved for me, as I can build now. Thanks!

**Comment by a4vi2r** [comment]
> run the gradle task "buildGameGenerationExecutableDebug"

How to do it? 
I can not found this task in Android Studio.

**Comment by ghelloe** [comment]
I still don't get it. I've been trying to figure these things for days. Help me pleaseee.
